server/src/main/java/com/sarny/spocone/server/game/ShipPlacementValidator.java

```diff
@@ -1,8 +1,125 @@
 package com.sarny.spocone.server.game;
 
+import java.util.HashMap;
+import java.util.Map;
+import java.util.Set;
+
 /**
  * @author Wojciech Makiela
  */
 class ShipPlacementValidator {
 
+    private final Board board;
+    private final ShipNeighbouringFieldsGenerator generator;
+    private final Map<Integer, Integer> shipsOfLengthToPlace;
+
+    private ShipPlacementValidator(HashMap<Integer, Integer> shipsOfLengthToPlace, Board board, ShipNeighbouringFieldsGenerator generator) {
+        this.shipsOfLengthToPlace = shipsOfLengthToPlace;
+        this.board = board;
+        this.generator = generator;
+    }
+
+    boolean allShipsPlaced() {
+        for (Integer leftToPlace : shipsOfLengthToPlace.values()) {
+            if (leftToPlace > 0) {
+                return false;
+            }
+        }
+        return true;
+    }
+
+    boolean validate(Ship ship) {
+        int shipsToPlace = shipsOfLengthToPlace.get(ship.length());
+        if (shipsToPlace <= 0) {
+            return false;
+        }
+        Set<Integer> fieldsAroundShip = generator.generateNeighbours(ship);
+        fieldsAroundShip.addAll(ship.toHit);
+
+        for (Integer field : fieldsAroundShip) {
+            if (board.getShipFromField(field) != null) {
+                return false;
+            }
+        }
+        return true;
+    }
+
+    void placedNewShip(Ship ship) {
+        int previousRemainingShips = shipsOfLengthToPlace.get(ship.length());
+        shipsOfLengthToPlace.put(ship.length(), previousRemainingShips - 1);
+    }
+
+    static class Builder implements WithShipsOfLength4, WithShipsOfLength3, WithShipsOfLength2, WithShipsOfLength1, WithGuaranteedMissGenerator, BuildShipPlacementValidator {
+
+        private int length4 = 1;
```

Comment:
length3 caused i was confused for few seconds. Change to lengthOfShip3 -> more readable